arangod/Aql/Executor/EnumerateNearVectorExecutor.cpp

```diff
@@ -0,0 +1,236 @@
+    // there are no results anymore for this input, so we can skip to next input
+    // row
+    if (_labels[_currentProcessedResultCount] == -1) {
+      _currentProcessedResultCount = _infos.getNumberOfResults();
+      break;
+    }
+    output.moveValueInto(
+        docOutId, _inputRow,
+        AqlValueHintUInt(_labels[_currentProcessedResultCount]));
+    output.moveValueInto(
+        distOutId, _inputRow,
+        AqlValueHintDouble(_distances[_currentProcessedResultCount]));
+    output.advanceRow();
+
+    ++_currentProcessedResultCount;
+  }
+
+  if (_currentProcessedResultCount == _infos.getNumberOfResults()) {
+    _resultsAreProcessed = true;
+  }
+}
+
+std::tuple<ExecutorState, NoStats, AqlCall>
+EnumerateNearVectorsExecutor::produceRows(AqlItemBlockInputRange& inputRange,
+                                          OutputAqlItemRow& output) {
+  LOG_INTERNAL << "hasDataRow: " << inputRange.hasDataRow();
+  while (!output.isFull()) {
+    if (!_initialized) {
+      if (!_inputRow.isInitialized()) {
+        fillInput(inputRange);
+      }
+
+      searchResults();
+    }
+    fillOutput(output);
+    if (_currentProcessedResultCount == _infos.getNumberOfResults()) {
+      _initialized = false;
+      break;
+    }
+  }
+
+  return {inputRange.upstreamState(), {}, output.getClientCall()};
+}
+
+[[nodiscard]] std::tuple<ExecutorState, Stats, size_t, AqlCall>
+EnumerateNearVectorsExecutor::skipRowsRange(AqlItemBlockInputRange& inputRange,
+                                            AqlCall& call) {
+  std::size_t skipped{0};
+  LOG_INTERNAL << "call: " << call << " fullCount: " << std::boolalpha
+               << call.fullCount << " needSkipMore: " << std::boolalpha
+               << call.needSkipMore()
+               << ", hasInputRange: " << inputRange.hasDataRow()
+               << ", resultsAreProcessed: " << _resultsAreProcessed;
+  while (call.needSkipMore()) {
+    // get an input row first, if necessary
+    if (!_inputRow.isInitialized() && !_initialized) {
+      std::tie(_state, _inputRow) = inputRange.peekDataRow();
+
+      if (inputRange.hasDataRow()) {
+        fillInput(inputRange);
+      } else {
+        break;
+      }
+    }
+
+    LOG_INTERNAL << "offset: " << call.getOffset()
+                 << ", _currentProcessedResultCount: "
+                 << _currentProcessedResultCount;
+    if (call.getOffset() == 0) {
+      TRI_ASSERT(call.needsFullCount())
+          << "if we need to skip and skip is 0 fullCount must be true";
+
+      auto fullCount =
+          _collection->count(&_trx, transaction::CountType::kNormal);
+      skipped = fullCount - _processedInputs * _infos.getNumberOfResults();
+      LOG_INTERNAL << "needsFullCount is set, fullCount: " << fullCount
+                   << " processed inputs: " << _processedInputs
+                   << " skipping now: " << skipped;
+      call.didSkip(skipped);
+      break;
+    }
```

Comment:
In principle (which means ignoring for a moment that we have to supply a finite `topK`), our total possible output should be 
```
total_number_of_input_rows * collection_count
```
, but is practically limited by `offset + limit`, so
```
min(offset + limit, total_number_of_input_rows * collection_count)
```
. When running into this `if` branch this means that `offset+limit` attains the minimum, and `input_rows*col_count` is at least as large. I think you're intention here is to skip the difference between the two values, but I don't think it's quite correct. For one, you didn't multiply the collection count with the number of input rows (so this part works iff there's exactly one input row).

Instead you can add `_collection->count()` for _each_ remaining (unprocessed) input row; in which case you also don't need to subtract anything.